### ISSUE TITLE
Support 'public' and 'private' keywords on types, fields, and methods

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#6067]
When we added initial support for `public` and `private` symbols to the language, we avoided adding support for two key categories:
- types (enums, classes, records, unions, type aliases)
- fields and methods

**Types**
For types, this was due to uncertainty about how to handle uses of the private types.  Should it be an error to define a public function which takes a private type as an argument, since there is no way for the client to obtain an instance of that type?  What about public functions that return instances of private types?  If returning instances of the private type is permissible, should calling any of its methods be permissible or should that be an error?  If returning instances of the private type is permissible, should the user be able to send that instance as an actual to another function?  Should a public class be allowed to inherit from a private class in the module in which it is defined?

My thinking is that it should be an error to define a public symbol that publicly uses a private type.  It will take effort to get the current implementation of function resolution to support this filtering, though.  There also might be issues with this approach that I haven't considered.  Some of the answers here will also need to be applied to `use` statements with `except` or `only` lists that fail to include a symbol relied upon by an imported symbol.

**Fields and Methods**
Fields and methods are another can of worms.  We have not had thorough discussion about the impact of inheritance on private fields and methods; other languages are known to support "protected" fields and methods, so we might consider adding that.  What about secondary methods?  Should they be allowed to call private methods defined in other modules?  Does private mean:
- private to the class or record?
- private to the module in which that symbol is defined?
- the union of the module in which it is defined and the class/record in which it is defined?
- or the intersection of those scopes?

Should fields be treated differently than methods in this regard?  Should secondary methods be treated differently than primary methods in this regard?

This issue will need to be thought about more carefully before any action can be taken on supporting `public` and `private` on fields and methods

**Associated Future Test(s):**
test/visibility/private/classes/*
test/visibility/private/typeAlias.chpl
test/visibility/private/privateEnum.chpl

**Comment by benharsh** [comment]
For secondary methods defined in other modules, I recommend we only allow usage of public fields/methods.

**Comment by mppf** [comment]
Requested by a user in #9099.

**Comment by bradcray** [comment]
I've added this to the Chapel 2.0 epic primarily because of its potential interaction with whether the default is public or private in prototype vs. non- module code.

**Comment by lydia-duncan** [comment]
@bradcray - does that mean you think it should be completed before Chapel 2.0, or in 2.0?

**Comment by bradcray** [comment]
One of those two.  

In more detail: Recall that we've discussed the idea of making declarations in prototype modules and standalone code be `public` by default ("I'm sketching code and want to be able to access everything unless I say otherwise") whereas non-prototype module declarations might be `private` by default ("Don't expose anything unless I intentionally requested it").  I don't think there's been enough discussion of this issue to know that it's our path forward, but I think it fits with the prototype vs. production theme well (issue #8641) and would strengthen the prototype module concept by having it affect multiple features of a module rather than just the current one (error-handling strictness).  That said, it would be a breaking change in the future once we made the switch (so we'd be reluctant to make it after 2.0).

**Comment by lydia-duncan** [comment]
I also agree that changing the default behavior of modules to `private` by default would be a breaking change

**Comment by bradcray** [comment]
I don't think I understand the distinction you're making between "before 2.0" and "for 2.0".  If 2.0 is the point at which we (attempt to) say "no more breaking changes after this" what's the difference?

**Comment by lydia-duncan** [comment]
I used "before 2.0" and "in 2.0".  I'm trying to distinguish "things that will need to happen before we can say 2.0" and "things that will happen in the 2.0 - 3.0 time frame, should a 3.0 be necessary".

**Comment by bradcray** [comment]
Oh, OK.  I was reading "in 2.0" as meaning "in the 2.0 release itself" which is why I was confused.  E.g., if constructors were deprecated in the 2.0 release and not before, I think that'd be OK.

**Comment by mppf** [comment]
Regarding fields and methods - I think that "private to the module in which that symbol is defined" is the right interpretation.

**Comment by lydia-duncan** [comment]
Note to self: there are some uses in the standard/internal/package modules that could potentially become private if we were able to make fields/methods private (depending on how we decide to handle private for secondary methods defined outside of the original module)

**Comment by lydia-duncan** [comment]
Note: the discussion on #23183 brought up a question about whether types that are only generic due to private fields should still be required to be declared with `(?)` in the contexts where that is required.  When designing private fields, we should decide on an answer to that question.

**Comment by jmag722** [comment]
Any update on the status of this feature? There was mention of it for the 2.0 roadmap, would it be deferred until another breaking change release (3.0)?

**Comment by bradcray** [comment]
Hi @jmag722 — This feature has not received any attention or discussion in recent years, I'm sorry to say, and as you're probably aware, did not get completed in time for Chapel 2.0.

It could be added and implemented anytime prior to Chapel 3.0 via the editions feature (see https://chapel-lang.org/blog/posts/announcing-chapel-2.5/#chapel-editions, https://chapel-lang.org/docs/main/technotes/editions.html, and https://chapel-lang.org/docs/main/developer/bestPractices/EditionsDev.html).  I also think that if we were to add these capabilities without changing the default public-ness/privacy of existing declarations that it could be considered a non-breaking change, in which case it could be added to the `main` branch / following release as an unstable feature once its design and implementation was complete.

Are you finding yourself in need of this feature?

**Comment by jmag722** [comment]
Hi @bradcray, it's more a question I guess of what's considered Chapel's modus operandi for encapsulation, and is that going to change in the near future. Workarounds I see are nested procedures or nested modules, where in the latter you control what gets imported to the outer scope. Maybe one isn't necessarily better than the other, but a simple `private` keyword on types/fields/methods feels natural. Defaulting to private has some benefits I guess, but that'd be a major break, and not one I feel strongly about. 

Also I haven't played with editions yet, something I will need to learn more about

**Comment by bradcray** [comment]
I agree that `private` feels like a natural extension to the language, and can go either way on the default-`private` from week to week ("productivity!" …  "safety!").  I suspect that in most cases it would be relatively easy to implement, though I think class hierarchies and generics may be two things that add complexity.

We have generally been shying away from implementing new language features that require compiler support in the resolution pass since completing the 2.0 release in order to focus on our [Dyno](https://chapel-lang.org/blog/tags/dyno/) compiler rework and avoid implementing the features twice—once in the current/"production" compiler and a second time in Dyno.  Our focus instead has been on trying to get Dyno feature complete and taking care of resolution by default and then implementing new features there—both to avoid the duplication and also because Dyno is much better architected and documented.  So features like this have unfortunately been on the back-burner as a result, unfortunately.

**Comment by lydia-duncan** [comment]
I think there's room for opinions on how the feature should look when it does get implemented, especially based on motivating use cases.  There's been many ideas toyed with over the years, but nothing serious as of yet and the dyno timeline is a limiting factor, as Brad mentioned.  But that means we aren't locked into anything, beyond avoiding an additional keyword reservation for "public" and "private" themselves.  We'll still have to think about interplay with the concepts that have been mentioned recently, and also forwarding.  But use cases would give us guidance on what sorts of pitfalls might lie ahead for users interested in the feature and where blindspots might lie both in the community and amongst the developers.
